acceptance/tests/environment/variables_refreshed_each_compilation.rb

```diff
@@ -15,62 +15,78 @@
   CONF
   # the module function loading logic is different from inside a single manifest
   #   we exercise both here
-  on(master, "mkdir -p #{fq_tmp_environmentpath}/modules/uptime/{manifests,functions}")
-  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/uptime/manifests/init.pp", <<-FILE)
-    class uptime {
-      notify { 'current uptime':
-        message => uptime::my_system_uptime()
+  on(master, "mkdir -p #{fq_tmp_environmentpath}/modules/custom_time/{manifests,functions,facts.d}")
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/custom_time/manifests/init.pp", <<-FILE)
+    class custom_time {
+      $t = custom_time::my_system_time()
+
+      notify { 'custom time':
+        message => "module_${t}_module",
       }
     }
   FILE
-  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/uptime/functions/my_system_uptime.pp", <<-FILE)
-    function uptime::my_system_uptime() {
-      $facts['system_uptime']
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/custom_time/functions/my_system_time.pp", <<-FILE)
+    function custom_time::my_system_time() {
+      $facts['custom_time']
     }
   FILE
   create_sitepp(master, tmp_environment, <<-SITE)
     function bar() {
-      $facts['system_uptime']['seconds']
+      $facts['custom_time']
     }
     class foo::bar {
       notify { "local_${bar()}_local": }
     }
     include foo::bar
-    include uptime
+    include custom_time
   SITE
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/custom_time/facts.d/custom_time.sh", <<-FILE)
+#!/bin/bash
+
+echo -n "custom_time=$(date +%s%N)"
+  FILE
+  create_remote_file(master, "#{fq_tmp_environmentpath}/modules/custom_time/facts.d/custom_time.ps1", <<-FILE)
```

Comment:
nit, it looks like this file gets created with permissions 777:

```
[root@j35a3o5f04bldcc facts.d]# ls -la
total 12
drwxrwxrwx.  2 root   root     49 Jan 18 15:52 .
drwxr-x---. 10 puppet puppet 4096 Jan 18 15:50 ..
-rwxrwxrwx.  1 root   root     54 Jan 18 15:52 custom_time.ps1
-rwxrwxrwx.  1 root   root     49 Jan 18 15:52 custom_time.sh
```

So unix agents will attempt to execute the ps1 external fact generating an error

```
Error: Facter: error while processing "/opt/puppetlabs/puppet/cache/facts.d/custom_time.ps1" for external facts: child process returned non-zero exit status (8).
```

It doesn't prevent the test from working, just was alarming when I first saw it.

Reply:
testing locally

Reply:
Okay, got this working without the error message:
```
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:31:43$ service puppetserver reload  
      Reloading puppetserver:       [  OK  ]                                            
                                                                                                                                           
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 4.46 seconds
                                                                                                                                           
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:31:47$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'                                    
      Info: Retrieving pluginfacts                                                                          
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d]/mode: mode changed '0755' to '0777'                                
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d/custom_time.ps1]/ensure: defined content as '{md5}92f6d3b2ced31367a12ba631a1204fe9'
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d/custom_time.sh]/ensure: defined content as '{md5}692ef0fe3bfe1c61d3f20726ff984e76'                                                         
      Info: Retrieving plugin                                                           
      Notice: /File[/opt/puppetlabs/puppet/cache/lib/facter]/ensure: removed                                                                                                                                
      Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet]/ensure: removed                 
      Info: Retrieving locales                                                                                                              
      Notice: /File[/opt/puppetlabs/puppet/cache/locales/ja]/ensure: removed                                                               
      Info: Caching catalog for qvbvzvxyvgrmnoe.delivery.puppetlabs.net                                                                               
      Info: Applying configuration version '1516401111'                                                                                                                                                     
      Notice: local_1516401110492982929_local                                                                                                                                                               
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401110492982929_local]/message: defined 'message' as 'local_1516401110492982929_local'            
      Notice: module_1516401110492982929_module                                                         
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401110492982929_module'                                                                                
      Notice: Applied catalog in 0.08 seconds                                                                                                   
                                                                                                                             
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 4.57 seconds                                       
    Exited: 2                                                                                                                                                                                               
                                                                                                                                                                            
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:31:52$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'                                   
      Info: Retrieving pluginfacts                                                                                                                                                                  
      Info: Retrieving plugin                                                           
      Info: Retrieving locales                                                                                                                                                                              
      Info: Caching catalog for qvbvzvxyvgrmnoe.delivery.puppetlabs.net                                                                                
      Info: Applying configuration version '1516401111'                                 
      Notice: local_1516401113871813704_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401113871813704_local]/message: defined 'message' as 'local_1516401113871813704_local'
      Notice: module_1516401113871813704_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401113871813704_module'
      Notice: Applied catalog in 0.05 seconds
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 2.40 seconds
    Exited: 2
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:31:55$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for qvbvzvxyvgrmnoe.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_1516401117255063700_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401117255063700_local]/message: defined 'message' as 'local_1516401117255063700_local'
      Notice: module_1516401117255063700_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401117255063700_module'
      Notice: Applied catalog in 0.05 seconds
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 2.38 seconds
    Exited: 2
    
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) 14:31:57$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d]/mode: mode changed '0755' to '0777'
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d/custom_time.ps1]/ensure: defined content as '{md5}92f6d3b2ced31367a12ba631a1204fe9'
      Notice: /File[/opt/puppetlabs/puppet/cache/facts.d/custom_time.sh]/ensure: defined content as '{md5}692ef0fe3bfe1c61d3f20726ff984e76'
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for xb36nhkgfs0s75s.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_1516401119473341611_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401119473341611_local]/message: defined 'message' as 'local_1516401119473341611_local'
      Notice: module_1516401119473341611_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401119473341611_module'
      Notice: Applied catalog in 0.03 seconds
    
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) executed in 2.24 seconds
    Exited: 2
    
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) 14:32:00$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for xb36nhkgfs0s75s.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_1516401121680212850_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401121680212850_local]/message: defined 'message' as 'local_1516401121680212850_local'
      Notice: module_1516401121680212850_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401121680212850_module'
      Notice: Applied catalog in 0.03 seconds
    
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) executed in 2.18 seconds
    Exited: 2
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) 14:32:03$ puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f356xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for xb36nhkgfs0s75s.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_1516401124783862339_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_1516401124783862339_local]/message: defined 'message' as 'local_1516401124783862339_local'
      Notice: module_1516401124783862339_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_1516401124783862339_module'
      Notice: Applied catalog in 0.02 seconds
    
    xb36nhkgfs0s75s.delivery.puppetlabs.net (ubuntu1404-64-1) executed in 2.01 seconds
    Exited: 2
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) 14:32:05$ cmd.exe /c puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f3
56xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Notice: /File[C:/ProgramData/PuppetLabs/puppet/cache/facts.d/custom_time.ps1]/ensure: defined content as '{md5}92f6d3b2ced31367a12ba631a1204fe9'
      Notice: /File[C:/ProgramData/PuppetLabs/puppet/cache/facts.d/custom_time.sh]/ensure: defined content as '{md5}692ef0fe3bfe1c61d3f20726ff984e76'
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for arvt9a7q25ubnp0.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_143213464971_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_143213464971_local]/message: defined 'message' as 'local_143213464971_local'
      Notice: module_143213464971_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_143213464971_module'
      Notice: Applied catalog in 0.04 seconds
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) executed in 10.58 seconds
    Exited: 2
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) 14:32:15$ cmd.exe /c puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f3
56xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for arvt9a7q25ubnp0.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_143223932590_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_143223932590_local]/message: defined 'message' as 'local_143223932590_local'
      Notice: module_143223932590_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_143223932590_module'
      Notice: Applied catalog in 0.04 seconds
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) executed in 10.23 seconds
    Exited: 2
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) 14:32:27$ cmd.exe /c puppet agent -t --server qvbvzvxyvgrmnoe.delivery.puppetlabs.net --environment variables_refreshed_each_compilation_f3
56xvpr
      Info: Using configured environment 'variables_refreshed_each_compilation_f356xvpr'
      Info: Retrieving pluginfacts
      Info: Retrieving plugin
      Info: Retrieving locales
      Info: Caching catalog for arvt9a7q25ubnp0.delivery.puppetlabs.net
      Info: Applying configuration version '1516401111'
      Notice: local_143235351810_local
      Notice: /Stage[main]/Foo::Bar/Notify[local_143235351810_local]/message: defined 'message' as 'local_143235351810_local'
      Notice: module_143235351810_module
      Notice: /Stage[main]/Custom_time/Notify[custom time]/message: defined 'message' as 'module_143235351810_module'
      Notice: Applied catalog in 0.04 seconds
    
    arvt9a7q25ubnp0.delivery.puppetlabs.net (windows2008r2-64-1) executed in 10.39 seconds
    Exited: 2
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:32:37$ puppet master --configprint config
      /etc/puppetlabs/puppet/puppet.conf
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 0.90 seconds
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:32:38$ if [ -f '/tmp/variables_refreshed_each_compilation.rb.9dDBeu/puppet.conf.bak' ]; then cat '/tmp/variables_refreshed_each_compilation.rb
.9dDBeu/puppet.conf.bak' > '/etc/puppetlabs/puppet/puppet.conf'; rm -f '/tmp/variables_refreshed_each_compilation.rb.9dDBeu/puppet.conf.bak'; fi
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 0.12 seconds
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:32:38$ service puppetserver reload
      Reloading puppetserver:       [  OK  ]
    
    qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 4.29 seconds
Begin teardown

qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) 14:32:42$ rm -rf /etc/puppetlabs/code/environments/variables_refreshed_each_compilation_f356xvpr

qvbvzvxyvgrmnoe.delivery.puppetlabs.net (redhat6-64-1) executed in 0.11 seconds
End teardown
ruby/puppet/acceptance/tests/environment/variables_refreshed_each_compilation.rb passed in 68.89 seconds
```